server/config/server.ts

```diff
@@ -0,0 +1,206 @@
+  }
+  return ['true', 'yes', 'y'].includes(val.toLowerCase().trim())
+}
+
+const envConfig: ServerConfig = {
+  database: {
+    sql: process.env.RCTF_DATABASE_URL ?? {
+      host: process.env.RCTF_DATABASE_HOST,
+      port: parseInt(process.env.RCTF_DATABASE_PORT) || undefined,
+      user: process.env.RCTF_DATABASE_USERNAME,
+      password: process.env.RCTF_DATABASE_PASSWORD,
+      database: process.env.RCTF_DATABASE_DATABASE
+    },
+    redis: process.env.RCTF_REDIS_URL ?? {
+      host: process.env.RCTF_REDIS_HOST,
+      port: parseInt(process.env.RCTF_REDIS_PORT) || undefined,
+      password: process.env.RCTF_REDIS_PASSWORD,
+      database: parseInt(process.env.RCTF_REDIS_DATABASE) || undefined
+    },
+    migrate: process.env.RCTF_DATABASE_MIGRATE
+  },
+  instanceType: process.env.RCTF_INSTANCE_TYPE,
+  tokenKey: process.env.RCTF_TOKEN_KEY,
+  origin: process.env.RCTF_ORIGIN,
+  ctftime: {
+    clientId: process.env.RCTF_CTFTIME_CLIENT_ID,
+    clientSecret: process.env.RCTF_CTFTIME_CLIENT_SECRET
+  },
+  userMembers: parseBoolEnv(process.env.RCTF_USER_MEMBERS),
+  sponsors: undefined,
+  homeContent: process.env.RCTF_HOME_CONTENT,
+  ctfName: process.env.RCTF_NAME,
+  meta: {
+    description: process.env.RCTF_META_DESCRIPTION,
+    imageUrl: process.env.RCTF_IMAGE_URL
+  },
+  logoUrl: process.env.RCTF_LOGO_URL,
+  globalSiteTag: process.env.RCTF_GLOBAL_SITE_TAG,
+  challengeProvider: {
+    name: undefined,
+    options: undefined
+  },
+  uploadProvider: {
+    name: undefined,
+    options: undefined
+  },
+  email: {
+    provider: {
+      name: undefined,
+      options: undefined
+    },
+    from: process.env.RCTF_EMAIL_FROM
+  },
+  divisions: undefined,
+  startTime: parseInt(process.env.RCTF_START_TIME) || undefined,
+  endTime: parseInt(process.env.RCTF_END_TIME) || undefined,
+  leaderboard: {
+    maxLimit: parseInt(process.env.RCTF_LEADERBOARD_MAX_LIMIT) || undefined,
+    maxOffset: parseInt(process.env.RCTF_LEADERBOARD_MAX_OFFSET) || undefined,
+    updateInterval: parseInt(process.env.RCTF_LEADERBOARD_UPDATE_INTERVAL) || undefined,
+    graphMaxTeams: parseInt(process.env.RCTF_LEADERBOARD_GRAPH_MAX_TEAMS) || undefined,
+    graphSampleTime: parseInt(process.env.RCTF_LEADERBOARD_GRAPH_SAMPLE_TIME) || undefined
+  },
+  loginTimeout: parseInt(process.env.RCTF_LOGIN_TIMEOUT) || undefined
+}
+
+const defaultConfig: PartialDeep<ServerConfig> = {
+  database: {
+    migrate: 'never'
+  },
+  instanceType: 'all',
+  userMembers: true,
+  sponsors: [],
+  homeContent: '',
+  challengeProvider: {
+    name: 'challenges/database'
+  },
+  uploadProvider: {
+    name: 'uploads/local'
+  },
+  meta: {
+    description: '',
+    imageUrl: ''
+  },
+  leaderboard: {
+    maxLimit: 100,
+    maxOffset: 4294967296,
+    updateInterval: 10000,
+    graphMaxTeams: 10,
+    graphSampleTime: 1800000
+  },
+  loginTimeout: 3600000
+}
+
+const _removeUndefined = <T>(o: Record<string, T>): Record<string, T> | undefined => {
```

Comment:
These utility functions should be defined in a separate file so they can be unit-tested without executing config loading code.